cardano-api/src/Cardano/Api/IPC/Monad.hs

```diff
@@ -52,44 +59,56 @@ executeLocalStateQueryExpr
+  => e `CouldBe` AcquiringFailure
+  => LocalNodeConnectInfo mode
+  -> Maybe ChainPoint
+  -> ExceptT (Variant e) (LocalStateQueryExpr (BlockInMode mode) ChainPoint (QueryInMode mode) () IO) a
+  -> ExceptT (Variant e) IO a
+executeLocalStateQueryExpr_ connectInfo mpoint f = do
+  tmvResultLocalState <- liftIO (newEmptyTMVarIO @(Either (Variant e) a))
   let waitResult = readTMVar tmvResultLocalState
 
-  connectToLocalNodeWithVersion
+  liftIO $ connectToLocalNodeWithVersion
     connectInfo
     (\ntcVersion ->
       LocalNodeClientProtocols
       { localChainSyncClient    = NoLocalChainSyncClient
-      , localStateQueryClient   = Just $ setupLocalStateQueryExpr waitResult mpoint tmvResultLocalState ntcVersion f
+      , localStateQueryClient   = Just $ setupLocalStateQueryExpr_ waitResult mpoint tmvResultLocalState ntcVersion f
       , localTxSubmissionClient = Nothing
       , localTxMonitoringClient = Nothing
       }
     )
 
-  first toAcquiringFailure <$> atomically waitResult
+  ExceptT . return =<< liftIO (atomically waitResult)
 
 -- | Use 'queryExpr' in a do block to construct monadic local state queries.
-setupLocalStateQueryExpr ::
-     STM x
+setupLocalStateQueryExpr_ :: ()
+  => e `CouldBe` AcquiringFailure
+  => STM x
      -- ^ An STM expression that only returns when all protocols are complete.
      -- Protocols must wait until 'waitDone' returns because premature exit will
      -- cause other incomplete protocols to abort which may lead to deadlock.
   -> Maybe ChainPoint
-  -> TMVar (Either Net.Query.AcquireFailure a)
+  -> TMVar (Either (Variant e) a)
   -> NodeToClientVersion
-  -> LocalStateQueryExpr (BlockInMode mode) ChainPoint (QueryInMode mode) () IO a
+  -> ExceptT (Variant e) (LocalStateQueryExpr (BlockInMode mode) ChainPoint (QueryInMode mode) () IO) a
   -> Net.Query.LocalStateQueryClient (BlockInMode mode) ChainPoint (QueryInMode mode) IO ()
-setupLocalStateQueryExpr waitDone mPointVar' resultVar' ntcVersion f =
+setupLocalStateQueryExpr_ waitDone mPointVar' resultVar' ntcVersion f =
   LocalStateQueryClient . pure . Net.Query.SendMsgAcquire mPointVar' $
     Net.Query.ClientStAcquiring
-    { Net.Query.recvMsgAcquired = runContT (runReaderT (runLocalStateQueryExpr f) ntcVersion) $ \result -> do
-        atomically $ putTMVar resultVar' (Right result)
+    { Net.Query.recvMsgAcquired = runContT (runReaderT (runLocalStateQueryExpr (runExceptT f)) ntcVersion) $ \result -> do
+        atomically $ putTMVar resultVar' result
         void $ atomically waitDone -- Wait for all protocols to complete before exiting.
         pure $ Net.Query.SendMsgRelease $ pure $ Net.Query.SendMsgDone ()
 
     , Net.Query.recvMsgFailure = \failure -> do
-        atomically $ putTMVar resultVar' (Left failure)
+        atomically $ putTMVar resultVar' (Left (DV.throw (toAcquiringFailure failure)))
         void $ atomically waitDone -- Wait for all protocols to complete before exiting.
         pure $ Net.Query.SendMsgDone ()
```

Comment:
This function is rewritten to `oops` without preserving a wrapper function because it is not exported and no-one calls it.  If we need one in the future, it is easy to write.